### PR TITLE
rust_g and CentCom Galactic Ban Database

### PR DESCRIPTION
**Pull request by ike709**
Ports https://github.com/BeeStation/BeeStation-Hornet/pull/2207

Admins will now be able to look up a player's bans from several other servers via the player panel.

My hope is that porting this to as many servers as possible will encourage more servers to make their bans publicly viewable so they can be included in this system. Direct access to a server's database is not required (or even supported).

Supported servers: 
* BeeStation 
* /vg/station
* OracleStation
* FTL13
* Fulpstation
* TGMC

Planned support that bobbah is working on:
* World Server
* Yogstation
* Any other server willing to make their bans publicly visible.

API: https://centcom.melonmesa.com
Source: https://github.com/bobbahbrown/CentCom

<hr>

This implementation differs slightly due to the differences between Halo: SSE and modern codebases. Also, this adds rust_g since you didn't already have it. It offloads a bunch of neat, expensive things from BYOND without the reverse engineering of Extools. More details here: https://github.com/tgstation/rust-g

**This PR has NOT been tested on Linux. Please test it on Linux since you use it for hosting.** The windows version works fine, though, so I don't expect anything to horrifically break.

<hr>

:cl: ike709 and bobbahbrown
rscadd: Admins can now see your bans on (some) other servers.
/:cl:

### CODE REVIEW

**Comment by ike709** [comment]
image

File: code/modules/admin/admin.dm [hunk 1]
```diff
@@ -1171,7 +1177,7 @@ var/global/floorIsLava = 0
 		out += "<b>Autotraitor <a href='?src=\ref[ticker.mode];toggle=autotraitor'>disabled</a></b>.<br/>"
 
 	out += "<b>All antag ids:</b>"
-	if(ticker.mode.antag_templates && ticker.mode.antag_templates.len).
+	if(ticker.mode.antag_templates && ticker.mode.antag_templates.len)
```
Comment (ike709) [inline code/modules/admin/admin.dm:1180]: Free fix courtesy of DreamChecker.

File: code/_helpers/unsorted.dm [hunk 2]
```diff
@@ -1142,3 +1142,5 @@ GLOBAL_DATUM_INIT(dview_mob, /mob/dview, new)
 
 /proc/pass()
 	return
+
+#define UNTIL(X) while(!(X)) sleep(world.tick_lag)
```
Comment (ike709) [inline code/_helpers/unsorted.dm:1146]: Normally this uses `stop_lag()` but you guys don't have it and porting it would be quite a bit more involved.

**Review by bloxgate — COMMENTED** [review]
We're not 100% on the rust-g addition at the moment, but in the meantime could you move the shared object into `lib/` so it's consistent with existing stuff?

File: code/modules/client/client_procs.dm [hunk 3]
```diff
@@ -78,6 +78,61 @@
...
+	else if(href_list["centcomlookup"])
+		if(!check_rights(R_ADMIN))
+			return
+
+		if(!config.centcom_ban_db)
+			to_chat(usr, "<span class='warning'>Centcom Galactic Ban DB is disabled!</span>")
+			return
+
+		//var/ckey = href_list["centcomlookup"]
+		var/ckey = "patriotwasp"
```
Comment (ike709) [inline code/modules/client/client_procs.dm]: Woops I left the test ckey in.

**Comment by ike709** [comment]
rust-g is used because it has more robust HTTP support than `world.Export()`.

Also if you build it yourself with all of the features enabled, it provides other benefits like improved file I/O for logging, etc.

**Comment by bloxgate** [comment]
My main problem with adding rust-g is that we're already not using `world.Export()` except as a fallback. Unless we do end up using the other features, rust-g only adds an async version of existing methods and a few request types we don't have much use for. The final decision on that part isn't mine however.

As for the rest of the PR, I've requested a review from the appropriate people.

**Comment by ike709** [comment]
Fulpstation bans now supported.

**Comment by ike709** [comment]
TGMC bans now supported.

**Comment by caelaislinn** [comment]
I can't see where in this PR our ban info gets uploaded, and what is the server URL for the centcom db that goes in our config file?

File: config/example/config.txt [hunk 4]
```diff
@@ -405,8 +405,12 @@ STARLIGHT 0
...
 
-## Clients will be unable to connect unless their build is equal to or higher than this (a number, e.g. 1000)	
-#MINIMUM_BYOND_BUILD
\ No newline at end of file
+## Clients will be unable to connect unless their build is equal to or higher than this (a number, e.g. 1000)
+#MINIMUM_BYOND_BUILD
+
+## Uncomment to enable global ban DB using the provided URL. The API should expect to receive a ckey at the end of the URL.
+## More API details can be found here: https://centcom.melonmesa.com
+CENTCOM_BAN_DB https://centcom.melonmesa.com/ban/search
```
Comment (ike709) [inline config/example/config.txt:416]: @caelaislinn ^

The config to enable it also contains the URL to use.

**Comment by ike709** [comment]
This doesn't upload bans, this is just a way of viewing bans that have been uploaded.

Directly accessing every server's DB isn't very cool and also has a lot of other problems, so instead bans are added to the system by querying an API on each server's webserver, which you'll have to setup separately.

Halo bans won't be included until that happens, but they don't have to be included to see other servers' bans.

**Comment by caelaislinn** [comment]
Yes I'm aware only a limited subset of ban info would be accessible via a separate db. We don't have the devs to commit to working on our own upload process, as this is an extremely low priority for me and no-one else is particularly enthusiastic either. 

I don't dislike it though and the admin team have approved it so I'm going to merge this.

**Review by caelaislinn — APPROVED** [review]

**Review by BDpuffy420 — APPROVED** [review]

**Pull request merged**